pkg/asset/cluster/cluster.go

```diff
@@ -129,12 +115,13 @@ func (c *Cluster) Generate(parents asset.Parents) (err error) {
 	}
 
 	for _, stage := range stages {
-		outputs, err := c.applyStage(platform, stage, terraformDirPath, tfvarsFiles)
+		outputs, stateFile, err := stage.Provision(tfvarsFiles, c.FileList) //TODO: make sure c.FileList ends up being used
 		if err != nil {
-			return errors.Wrapf(err, "failure applying terraform for %q stage", stage.Name())
+			errors.Wrapf(err, "provisioning infrastructure in stage %s", stage.Name())
 		}
 		tfvarsFiles = append(tfvarsFiles, outputs)
 		c.FileList = append(c.FileList, outputs)
+		c.FileList = append(c.FileList, stateFile)
```

Comment:
1. Handle nil values like https://github.com/openshift/installer/pull/7534/commits/955a030c528a4b8237d916b6cd98e79bd2961cf6
2. Further consider the expectations around `outputs` & `statefile`